### PR TITLE
[FLINK-13101][datastream] Introduce blockingConnectionsBetweenChains property of StreamGraph

### PR DESCRIPTION
**Pull request by ifndef-SleePy**
## What is the purpose of the change

* **This is based on https://github.com/apache/flink/pull/8982**

* The reason of introducing it is to satisfy the requirement of Blink batch planner.
* Because the current scheduling strategy is a bit simple. It can not support some complex scenarios, like a batch job with resources limited.
* To be honest, it's probably a work-around solution. However it's an internal implementation, we can replace it when we are able to support batch job by scheduling strategy.

## Brief change log

* Add a property `blockingConnectionsBetweenChains` of `StreamGraph`
* The property `blockingConnectionsBetweenChains` means, if there are some stream edges that can not be chained and the shuffle mode of edge is not specified, translate these edges into `BLOCKING` result partition type.

## Verifying this change

* This change added unit tests

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? **yes**
  - If yes, how is the feature documented? **JavaDocs**

### CODE REVIEW

**Comment by ifndef-SleePy** [comment]
@flinkbot attention @aljoscha

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraph.java [hunk 1]
```diff
@@ -97,6 +97,8 @@
 
 	private TimeCharacteristic timeCharacteristic;
 
+	private boolean blockingAfterChainingOff;
```
Comment (guoweiM) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraph.java]: I think `slotSharingEnableAfterChain` is more accurate?  From the discussion in the slack what my understand is that the planner only controls whether slot is sharing and ShuffleMode is determined by this currently.
What do you think?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphGenerator.java [hunk 2]
```diff
@@ -123,6 +123,12 @@
 
 	private String jobName = DEFAULT_JOB_NAME;
 
+	/**
+	 * If there are some stream edges that can not be chained and the shuffle mode of edge is not specified,
+	 * translate these edges into `BLOCKING` result partition type.
+	 */
+	private boolean blockingAfterChainingOff = false;
```
Comment (guoweiM) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphGenerator.java]: Why add this property?  Planner can get `StreamGraph` and set the property to it.

Comment (ifndef-SleePy) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphGenerator.java]: Yes, planner could set `StreamGraph` directly. But that's a work-around way for 1.9. Actually the `StreamGraphGenerator` should be the only entry of `StreamGraph`. So just keep it here.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraph.java (hunk 1 above)
Comment (ifndef-SleePy) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraph.java]: I'm not sure we have reached an agreement of details. I guess we have a divergence here.
Let's wait for the opinion of Aljoscha and Stephan.

**Comment by ifndef-SleePy** [comment]
@flinkbot attention @aljoscha

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraph.java (hunk 1 above)
Comment (guoweiM) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraph.java]: ok. I just remind another thing that Blink Planner might set "slot-sharing after chaining" off and ask ShuffleMode between tasks to be PIPELINE.

**Comment by ifndef-SleePy** [comment]
I have implemented an alternative solution https://github.com/apache/flink/pull/9000. Maybe the other one is more close with the original discussion.

**Comment by zentol** [comment]
CI report for commit 34336fc5898ab8bd74eb30153dcc13dd9611983f: SUCCESS [Travis](https://travis-ci.org/dataArtisans/flink/builds/554657817?utm_source=github_status&utm_medium=notification)

**Comment by aljoscha** [comment]
Merged. 👌

**Pull request closed without merging**